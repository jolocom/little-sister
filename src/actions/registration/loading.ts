import {AnyAction} from 'redux'

export const setLoadingMsg = (loadingMsg: string) => {
  return {
    type: 'SET_LOADING_MSG',
    value: loadingMsg
  }
}

export const startLoading = (loading: boolean) => {
  return {
    type: 'START_LOADING',
    value: loading
  }
}

export const finishLoading = (loading: boolean) => {
  return {
    type: 'FINISH_LOADING',
    value: loading
  }
}

export const generateAndEncryptKeyPairs = () => {
  return (dispatch: (actions: AnyAction) => void, {backendMiddleware}) => {
    const randomString = '13912643311766764847120568039921' // TODO: grab from the state
    const password = 'Password1' // TODO: grab from the state

    dispatch(setLoadingMsg('Generating keys'))

    const {
      didDocument,
      mnemonic,
      masterKeyWIF,
      genericSigningKeyWIF,
      ethereumKeyWIF
    } = backendMiddleware.jolocomLib.identity.create(randomString)

    dispatch(setLoadingMsg('Fueling with Ether'))
  }
}
