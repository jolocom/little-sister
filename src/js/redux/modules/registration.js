import * as _ from 'lodash'
import Immutable from 'immutable'
import { makeActions } from './'
import * as router from './router'
import { isPasswordValid, checkPassStrength,
  passwordCharacters } from '../../lib/password-util'

const NEXT_ROUTES = {
  '/registration': '/registration/entropy',
  '/registration/entropy': '/registration/user-type',
  '/registration/user-type': '/registration/phrase-info',
  '/registration/phrase-info': '/registration/email',
  '/registration/email': '/registration/password',
  '/registration/password': '/registration/pin',
  '/registration/write-phrase': '/registration/pin'
}

const CHECK_BEFORE_SWITCHING = {
  '/registration': ['username', 'valid'],
  '/registration/entropy': ['passphrase', 'sufficientEntropy'],
  '/registration/user-type': ['userType', 'valid'],
  '/registration/write-phrase': ['passphrase', 'writtenDown'],
  '/registration/phrase-info': ['passphrase', 'writtenDown'],
  '/registration/email': ['email', 'valid'],
  '/registration/password': ['password', 'valid'],
  '/registration/pin': ['pin', 'valid']
}

const actions = module.exports = makeActions('registration', {
  goForward: {
    expectedParams: [],
    creator: () => {
      return (dispatch, getState) => {
        const state = getState()
        if (state.getIn(['registration', 'complete'])) {
          dispatch(actions.registerWallet())
        } else {
          const nextUrl = helpers._getNextURLFromState(state)
          dispatch(router.pushRoute(nextUrl))
        }
      }
    }
  },
  setUserType: {
    expectedParams: ['value']
  },
  setMaskedImageUncovering: {
    expectedParams: ['value']
  },
  addEntropyFromDeltas: {
    expectedParams: ['x', 'y'],
    creator: (params) => {
      return (dispatch, getState, {backend, services}) => {
        if (getState().getIn(
          ['registration', 'passphrase', 'phrase']
        )) {
          return
        }

        const entropy = services.entropy
        entropy.addFromDelta(params.x)
        entropy.addFromDelta(params.y)
        if (params.dz) {
          entropy.addFromDelta(params.z)
        }

        dispatch(actions.setEntropyStatus.buildAction({
          sufficientEntropy: entropy.getProgress() >= 1,
          progress: entropy.getProgress()
        }))

        if (entropy.isReady()) {
          const randomString = entropy.getRandomString(12)
          dispatch(actions.setPassphrase(
            backend.wallet.generateSeedPhrase(randomString)
          ))
        }
      }
    }
  },
  setEntropyStatus: {
    expectedParams: ['sufficientEntropy', 'progress']
  },
  setPassphrase: {
    expectedParams: ['phrase']
  },
  setRandomString: {
    expectedParams: ['randomString']
  },
  setPassphraseWrittenDown: {
    expectedParams: ['value']
  },
  switchToExpertMode: {
    expectedParams: []
  },
  setPin: {
    expectedParams: ['value']
  },
  setPinConfirm: {
    expectedParams: ['value']
  },
  setPinFocused: {
    expectedParams: ['value']
  },
  submitPin: {
    expectedParams: [],
    creator: () => {
      return (dispatch, getState) => {
        const pinState = getState().getIn(['registration', 'pin'])
        if (!pinState.get('valid')) {
          return
        }
        dispatch(actions.goForward())
      }
    }
  },
  setUsername: {
    expectedParams: ['value']
  },
  setEmail: {
    expectedParams: ['value']
  },
  checkEmail: {
    expectedParams: [],
    creator: () => {
      return (dispatch, getState) => {
        const emailState = getState().getIn(['registration', 'email'])
        if (emailState.get('valid')) {
          dispatch(actions.goForward())
        } else {
          dispatch(actions.emailError())
        }
      }
    }
  },
  emailError: {
    expectedParams: []
  },
  setPassword: {
    expectedParams: ['value']
  },
  setRepeatedPassword: {
    expectedParams: ['value']
  },
  checkUsername: {
    expectedParams: [],
    async: true,
    creator: (params) => {
      return (dispatch, getState) => {
        const state = getState().get('registration').toJS()
        dispatch(actions.checkUsername.buildAction(params, (backend) => {
          return backend.accounts
            .checkUsername(state.username.value).then((params) => {
              dispatch(actions.goForward())
            })
        }))
      }
    }
  },
  registerWallet: {
    expectedParams: [],
    async: true,
    creator: (params) => {
      return (dispatch, getState, {services, backend}) => {
        const state = getState().get('registration').toJS()
        dispatch(actions.registerWallet.buildAction(params, async () => {
          const userType = state.userType.value
          if (userType === 'expert') {
            const { wallet } = await services.auth.registerWithSeedPhrase({
              userName: state.username.value,
              seedPhrase: state.passphrase.phrase,
              pin: state.pin.value
            })

            await backend.accounts.solidRegister(
              state.username.value,
              state.passphrase.phrase,
              wallet.webIDPrivateKey
            )

            await backend.accounts.solidLogin(
              state.username.value,
              state.passphrase.phrase,
              wallet.webIDPrivateKey
            )

            await backend.solid.setIdentityContractAddress(
              wallet.webId,
              wallet.identityAddress
            )

            dispatch(router.pushRoute('/wallet'))
            return
          } else {
            return services.auth.registerWithCredentials({
              userName: state.username.value,
              email: state.email.value,
              password: state.password.value,
              pin: state.pin.value
            }).then(({wallet}) => {
              return backend.accounts.solidRegister(state.username.value,
                state.passphrase.phrase, wallet.webIDPrivateKey)
            }).then(({webid}) => {
              dispatch(router.pushRoute('/wallet'))
              return
            })
          }
        })
      )
      }
    }
  }
})

const initialState = Immutable.fromJS({
  username: {
    value: '',
    checking: false,
    errorMsg: '',
    valid: false,
    alphaNum: false
  },
  Repuation: 0,
  email: {
    value: '',
    valid: false,
    errorMsg: ''
  },
  password: {
    value: '',
    repeated: '',
    strength: 'weak',
    hasLowerCase: false,
    hasUpperCase: false,
    hasDigit: false,
    valid: false
  },
  pin: {
    value: '',
    focused: false,
    confirm: false,
    valid: false
  },
  userType: {
    value: '',
    valid: false
  },
  maskedImage: {
    uncovering: false
  },
  passphrase: {
    sufficientEntropy: false,
    progress: 0,
    randomString: '',
    phrase: '',
    writtenDown: false,
    valid: false
  },
  wallet: {
    registering: false,
    registered: false,
    errorMsg: null
  },
  complete: false
})

module.exports.default = (state = initialState, action = {}) => {
  state = state.set('complete', helpers._isComplete(state))

  switch (action.type) {
    case actions.setUserType.id:
      const valid = ['expert', 'layman'].indexOf(action.value) !== -1
      if (action.value && !valid) {
        throw Error('Invalid user type: ' + action.value)
      }

      return state.merge({
        userType: {
          value: action.value,
          valid
        }
      })

    case actions.setPassword.id:
      const oldRepeatedValue = state.get('password').get('repeated')
      const validPassword = isPasswordValid(action.value, oldRepeatedValue)
      const characters = passwordCharacters(action.value)
      const passwordStrength = checkPassStrength(action.value)

      return state.mergeIn(
        ['password'],
        {
          value: action.value,
          repeated: '',
          valid: validPassword,
          hasLowerCase: characters.lowerCase,
          hasUpperCase: characters.upperCase,
          hasDigit: characters.digit,
          strength: passwordStrength
        }
      )
    case actions.setRepeatedPassword.id:
      const passwordValue = state.get('password').get('value')
      const validRepeatedValue = isPasswordValid(action.value, passwordValue)

      return state.mergeIn(['password'], {
        repeated: action.value,
        valid: validRepeatedValue
      })
    case actions.setEntropyStatus.id:
      return state.mergeDeep({
        passphrase: {
          sufficientEntropy: action.sufficientEntropy,
          progress: action.progress
        }
      })
    case actions.setRandomString.id:
      return state.mergeIn(['passphrase'], {
        randomString: action.randomString
      })
    case actions.setPassphrase.id:
      return state.mergeIn(['passphrase'], {
        phrase: action.phrase
      })
    case actions.setPin.id:
      if (!/^[0-9]{0,4}$/.test(action.value)) {
        return state
      }

      return state.mergeIn(['pin'], {
        value: action.value,
        valid: action.value.length === 4
      })
    case actions.setPinConfirm.id:
      return state.mergeDeep({
        pin: {
          confirm: action.value
        }
      })
    case actions.setPinFocused.id:
      return state.mergeDeep({
        pin: {
          focused: action.value
        }
      })
    case actions.setMaskedImageUncovering.id:
      return state.setIn(['maskedImage', 'uncovering'], action.value)

    case actions.setEmail.id:
      return state.mergeDeep({
        email: {
          value: action.value,
          valid: /^([\w.]+)@([\w.]+)\.(\w+)/.test(action.value),
          errorMsg: ''
        }
      })
    case actions.emailError.id:
      return state.mergeDeep({
        email: {
          errorMsg: 'This email address is invalid. Please check it again'
        }
      })
    case actions.setPassphraseWrittenDown.id:
      return state.mergeDeep({
        passphrase: {
          writtenDown: action.value,
          valid: !!state.getIn('passphrase', 'phrase') && action.value
        }
      })
    case actions.registerWallet.id:
      return state.mergeDeep({
        wallet: {
          registering: true,
          registered: false,
          errorMsg: null
        }
      })
    case actions.registerWallet.id_success:
      return state.mergeDeep({
        wallet: {
          registering: true,
          registered: true
        }
      })
    case actions.registerWallet.id_fail:
      return state.mergeDeep({
        wallet: {
          registering: false,
          registered: false,
          errorMsg: action.error.message
        }
      })

    case actions.setUsername.id:
      return state.mergeDeep({
        username: {
          value: action.value,
          alphaNum: (/^[a-z0-9]+$/i.test(action.value)),
          valid: action.value.trim() !== '',
          errorMsg: ''
        }
      })

    case actions.checkUsername.id:
      return state.mergeDeep({
        username: {
          checking: true
        }
      })
    case actions.checkUsername.id_success:
      return state.mergeDeep({
        username: {
          checking: false,
          errorMsg: ''
        }
      })
    case actions.checkUsername.id_fail:
      return state.mergeDeep({
        username: {
          checking: false,
          errorMsg: action.error.message
        }
      })
    default:
      return state
  }
}

const helpers = module.exports.helpers = {}
helpers._isComplete = (state) => {
  const isFieldValid = (fieldName) => state.getIn([fieldName, 'valid'])
  const areFieldsValid = (fields) => _.every(fields, isFieldValid)

  let complete = areFieldsValid(['username', 'userType', 'pin'])
  if (state.getIn(['userType', 'value']) === 'layman') {
    complete = complete && areFieldsValid(['email', 'password'])
  } else {
    complete = complete && areFieldsValid(['passphrase'])
  }

  return complete
}

helpers._getNextURLFromState = (state) => {
  const currentPath = state.get('routing').locationBeforeTransitions.pathname
  const userType = state.getIn(['registration', 'userType', 'value'])
  if (!helpers._canGoForward(state, currentPath)) {
    if ((currentPath === '/registration/write-phrase') &&
      (userType === 'layman')) {
      return '/registration/phrase-info'
    }
    if ((currentPath === '/registration/phrase-info') &&
      (userType === 'expert')) {
      return '/registration/write-phrase'
    }
    return null
  }
  return helpers._getNextURL(currentPath, userType)
}

helpers._getNextURL = (currentPath, userType) => {
  if (currentPath === '/registration/user-type') {
    return userType === 'expert'
              ? '/registration/write-phrase'
              : '/registration/phrase-info'
  }

  return NEXT_ROUTES[currentPath]
}

helpers._canGoForward = (state, currentPath) => {
  const toCheck = CHECK_BEFORE_SWITCHING[currentPath]
  let result = !toCheck || state.getIn(['registration', toCheck[0], toCheck[1]])
  return result || false
}
