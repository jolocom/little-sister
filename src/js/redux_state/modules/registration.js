import every from 'lodash/every'
import Immutable from 'immutable'
import { makeActions } from './'
import { deriveMasterKeyPair, deriveGenericSigningKeyPair } from 'redux_state/key-derivation'
import router from './router'
import Mnemonic from 'bitcore-mnemonic'

const NEXT_ROUTES = {
  '/registration': '/registration/entropy',
  '/registration/entropy': '/registration/write-phrase'
}

export const actions = makeActions('registration', {
  goForward: {
    expectedParams: [],
    creator: () => {
      return (dispatch, getState) => {
        const state = getState()
        if (state.getIn(['registration', 'complete'])) {
          dispatch(actions.registerWallet())
        } else {
          let nextURL = helpers._getNextURLfromState(state)
          dispatch(router.pushRoute(nextURL))
        }
      }
    }
  },
  setMaskedImageUncovering: {
    expectedParams: ['value']
  },
  addEntropyFromDeltas: {
    expectedParams: ['x', 'y'],
    creator: (params) => {
      return (dispatch, getState, {backend, services}) => {
        if (getState().getIn(
          ['registration', 'passphrase', 'phrase']
        )) {
          return
        }

        const entropy = services.entropy
        entropy.addFromDelta(params.x)
        entropy.addFromDelta(params.y)
        if (params.dz) {
          entropy.addFromDelta(params.z)
        }

        dispatch(actions.setEntropyStatus.buildAction({
          sufficientEntropy: entropy.getProgress() >= 1,
          progress: entropy.getProgress()
        }))

        if (entropy.isReady()) {
          const randomString = entropy.getRandomString(12)
          dispatch(actions.setRandomString(randomString))
        }
      }
    }
  },
  submitEntropy: {
    expectedParams: [],
    creator: () => {
      return (dispatch, getState) => {
        const entropyState = getState().getIn([
          'registration',
          'passphrase',
          'sufficientEntropy'
        ])

        if (entropyState) {
          return dispatch(actions.generateKeyPairs())
        }

        throw new Error('Not enough entropy!')
      }
    }
  },
  generateKeyPairs: {
    expectedParams: [],
    async: true,
    creator: (params) => {
      return (dispatch, getState, {services}) => {
        const randomStringState = getState().getIn([
          'registration',
          'passphrase',
          'randomString'
        ])

        if (randomStringState === '') {
          return
        }
        // eslint-disable-next-line max-len
        dispatch(actions.generateKeyPairs.buildAction(params, async () => {
          // why does this only function as an async?
          const entropy = services.entropy
          let seed = new Mnemonic(entropy.getHashedEntropy(randomStringState), Mnemonic.Words.ENGLISH)
          let masterKeyPair = deriveMasterKeyPair(seed)
          // TODO: Save masterKeyPair
          let genericSigningKey = deriveGenericSigningKeyPair(masterKeyPair)
          dispatch(actions.setPassphrase(seed.phrase))
          dispatch(actions.goForward())
        }))
      }
    }
  },

  setEntropyStatus: {
    expectedParams: ['sufficientEntropy', 'progress']
  },
  setPassphrase: {
    expectedParams: ['phrase']
  },
  setRandomString: {
    expectedParams: ['randomString']
  },
  setPassphraseWrittenDown: {
    expectedParams: ['value']
  },
  setUsername: {
    expectedParams: ['value']
  },
  setValueOwnURL: {
    expectedParams: ['value']
  },
  toggleHasOwnURL: {
    expectedParams: ['value']
  },
  checkCredentials: {
    expectedParams: [],
    async: true,
    creator: params => {
      return (dispatch, getState) => {
        const state = getState().get('registration').toJS()
        dispatch(actions.checkCredentials.buildAction(params, (backend) => {
          return backend.gateway
            .checkUserDoesNotExist({userName: state.username.value})
            .then(params => {
              if (state.ownURL.valueOwnURL.length > 1) {
                dispatch(actions.checkOwnUrl())
              } else {
                dispatch(actions.goForward())
              }
            })
        }))
      }
    }
  },
  checkOwnUrl: {
    expectedParams: [],
    async: true,
    creator: (params) => {
      return (dispatch, getState, {backend}) => {
        const state = getState().get('registration').toJS()
        dispatch(actions.checkOwnUrl.buildAction(params, (backend) => {
          return backend.gateway.checkOwnUrlDoesExist({
            userName: state.username.value,
            gatewayUrl: state.ownURL.valueOwnURL
          })
          .then((params) => {
            dispatch(actions.goForward())
          })
        }))
      }
    }
  },
  registerWallet: {
    expectedParams: [],
    async: true,
    creator: (params) => {
      return (dispatch, getState, {services, backend}) => {
        const state = getState().get('registration').toJS()
        dispatch(actions.registerWallet.buildAction(params, async () => {
          await services.auth.register({
            userName: state.username.value,
            seedPhrase: state.passphrase.phrase,
            inviteCode: state.inviteCode,
            gatewayUrl: state.ownURL.valueOwnURL
          })
          await services.auth.login({
            seedPhrase: state.passphrase.phrase,
            gatewayUrl: state.ownURL.valueOwnURL
          })
          dispatch(router.pushRoute('/wallet'))
        })
      )
      }
    }
  },
  setInviteCode: {
    expectedParams: ['value']
  }
})

const initialState = Immutable.fromJS({
  username: {
    value: '',
    checking: false,
    errorMsg: '',
    valid: false,
    alphaNum: false
  },
  ownURL: {
    hasOwnURL: false,
    errorMsg: '',
    valueOwnURL: ''
  },
  maskedImage: {
    uncovering: false
  },
  passphrase: {
    sufficientEntropy: false,
    progress: 0,
    randomString: '',
    phrase: '',
    generating: false,
    generated: false,
    writtenDown: false,
    valid: false
  },
  wallet: {
    registering: false,
    registered: false,
    errorMsg: null
  },
  inviteCode: null,
  complete: false
})

export default (state = initialState, action = {}) => {
  state = state.set('complete', helpers._isComplete(state))

  switch (action.type) {

    case actions.setEntropyStatus.id:
      return state.mergeDeep({
        passphrase: {
          sufficientEntropy: action.sufficientEntropy,
          progress: action.progress
        }
      })

    case actions.setRandomString.id:
      return state.mergeIn(['passphrase'], {
        randomString: action.randomString
      })

    case actions.setPassphrase.id:
      return state.mergeIn(['passphrase'], {
        phrase: action.phrase
      })

    case actions.generateKeyPairs.id:
      return state.mergeDeep({
        passphrase: {
          generating: true,
          generated: false,
          errorMsg: null
        }
      })

    case actions.generateKeyPairs.id_fail:
      return state.mergeDeep({
        passphrase: {
          generating: false,
          generated: false,
          errorMsg: 'generating passphrase has failed'
        }
      })

    case actions.generateKeyPairs.id_success:
      return state.mergeDeep({
        passphrase: {
          generating: false,
          generated: true,
          errorMsg: null
        }
      })

    case actions.setMaskedImageUncovering.id:
      return state.setIn(['maskedImage', 'uncovering'], action.value)

    case actions.setPassphraseWrittenDown.id:
      return state.mergeDeep({
        passphrase: {
          writtenDown: action.value,
          valid: !!state.getIn('passphrase', 'phrase') && action.value
        }
      })

    case actions.registerWallet.id:
      return state.mergeDeep({
        wallet: {
          registering: true,
          registered: false,
          errorMsg: null
        }
      })

    case actions.registerWallet.id_success:
      return state.mergeDeep({
        wallet: {
          registering: true,
          registered: true
        }
      })

    case actions.registerWallet.id_fail:
      return state.mergeDeep({
        wallet: {
          registering: false,
          registered: false,
          errorMsg: action.error.message
        }
      })

    case actions.setUsername.id:
      return state.mergeDeep({
        username: {
          value: action.value,
          alphaNum: (/^[a-z0-9]+$/i.test(action.value)),
          valid: action.value.trim() !== '',
          errorMsg: ''
        }
      })

    case actions.checkCredentials.id:
      return state.mergeDeep({
        username: {
          checking: true
        }
      })

    case actions.checkCredentials.id_success:
      return state.mergeDeep({
        username: {
          checking: false,
          errorMsg: ''
        }
      })

    case actions.checkCredentials.id_fail:
      return state.mergeDeep({
        username: {
          checking: false,
          errorMsg: action.error.message
        }
      })

    // case actions.setInviteCode.id:
    //   return state.merge({
    //     inviteCode: action.code
    //   })

    case actions.toggleHasOwnURL.id:
      return state.mergeIn(['ownURL'], {
        hasOwnURL: action.value
      })

    case actions.setValueOwnURL.id:
      return state.mergeIn(['ownURL'], {
        valueOwnURL: action.value
      })

    case actions.checkOwnUrl.id:
      return state.mergeIn(['ownURL'], {
        errorMsg: ''
      })

    case actions.checkOwnUrl.id_success:
      return state.mergeIn(['ownURL'], {
        errorMsg: ''
      })

    case actions.checkOwnUrl.id_fail:
      return state.mergeIn(['ownURL'], {
        errorMsg: action.error.message
      })

    default:
      return state
  }
}

export const helpers = {
  _isComplete: (state) => {
    const isFieldValid = (fieldName) => state.getIn([fieldName, 'valid'])
    const areFieldsValid = (fields) => every(fields, isFieldValid)
    return areFieldsValid(['username', 'passphrase'])
  },
  _getNextURLfromState: (state) => {
    const currentPath = state.get('routing').locationBeforeTransitions.pathname
    return NEXT_ROUTES[currentPath]
  }
}
