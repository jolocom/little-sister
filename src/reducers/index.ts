import { combineReducers } from 'redux'
import { accountReducer, AccountState } from 'src/reducers/account/'
import {
  registrationReducer,
  RegistrationState,
} from 'src/reducers/registration/'
import { navigationReducer } from 'src/reducers/navigation/'
import { NavigationState } from 'react-navigation'
import { ssoReducer, SsoState } from 'src/reducers/sso/'

export const rootReducer = combineReducers({
  account: accountReducer,
  registration: registrationReducer,
  navigation: navigationReducer,
  sso: ssoReducer,
})

export interface RootState {
  readonly account: AccountState
  readonly registration: RegistrationState
  readonly navigation: NavigationState
  readonly sso: SsoState
}
