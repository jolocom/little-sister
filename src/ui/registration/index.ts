export { Loading } from './containers/loading'
export { SeedPhrase } from './containers/seedPhrase'
