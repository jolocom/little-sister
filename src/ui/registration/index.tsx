export { Entropy } from './containers/entropy'