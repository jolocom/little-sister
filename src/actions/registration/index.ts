import { navigationActions } from 'src/actions/'
import { routeList } from 'src/routeList'
import * as loading from 'src/actions/registration/loadingStages'
import { setDid } from 'src/actions/account'
import { ThunkAction } from 'src/store'
import { navigatorResetHome } from '../navigation'
import { setAutoBackup, setSeedPhraseSaved } from '../recovery'
import { EncryptedData } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'

export const setLoadingMsg = (loadingMsg: string) => ({
  type: 'SET_LOADING_MSG',
  value: loadingMsg,
})

export const setIsRegistering = (value: boolean) => ({
  type: 'SET_IS_REGISTERING',
  value,
})

export const createIdentity = (encodedEntropy: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  dispatch(
    navigationActions.navigate({
      routeName: routeList.RegistrationProgress,
    }),
  )

  const isRegistering = getState().registration.loading.isRegistering
  if (isRegistering) return

  dispatch(setIsRegistering(true))

  dispatch(setLoadingMsg(loading.loadingStages[0]))
  await backendMiddleware.createKeyProvider(encodedEntropy)

  dispatch(setLoadingMsg(loading.loadingStages[1]))
  await backendMiddleware.fuelKeyWithEther()

  dispatch(setLoadingMsg(loading.loadingStages[2]))
  const identity = await backendMiddleware.createIdentity()

  dispatch(setDid(identity.did))
  dispatch(setLoadingMsg(loading.loadingStages[3]))
  dispatch(setIsRegistering(false))

  return dispatch(navigatorResetHome())
}

export const recoverFromSeedPhrase = (mnemonic: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  dispatch(setIsRegistering(true))
  let pubKey
  try {
    pubKey = await backendMiddleware.recoverKeyProvider(mnemonic)
  } catch (e) {
    return dispatch(setIsRegistering(false))
  }

  const backup = await backendMiddleware.fetchBackup()
  if (backup) {
    dispatch(setAutoBackup(true, false))
    return dispatch(recoverIdentity(backup))
  }

  // No backup found, user needs to import manually
  // public key is needed to validate if the imported backup belongs to the recovered identity
  return dispatch(
    navigationActions.navigate({
      routeName: routeList.ImportBackup,
      params: { pubKey },
    }),
  )
}

export const recoverIdentity = (
  encryptedBackup?: EncryptedData,
): ThunkAction => async (dispatch, getState, backendMiddleware) => {
  // the backup should be validated before to avoid decryption failures at this point
  const backup = encryptedBackup
    ? await backendMiddleware.decryptBackup(encryptedBackup)
    : undefined
  const identity = await backendMiddleware.recoverIdentity(
    backup ? backup.did : undefined,
  )
  if (backup) await backendMiddleware.recoverData(backup)

  await dispatch(setDid(identity.did))
  await dispatch(setSeedPhraseSaved())
  await dispatch(setIsRegistering(false))
  return dispatch(navigatorResetHome())
}
