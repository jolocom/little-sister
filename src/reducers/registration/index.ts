import { combineReducers } from 'redux'
import { loading } from 'src/reducers/registration/loading'

export interface LoadingState {
  readonly loadingMsg: string
}

export interface RegistrationState {
  readonly loading: LoadingState
}

export const registrationReducer = combineReducers({
  loading
})
