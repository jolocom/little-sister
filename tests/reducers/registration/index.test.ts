import { registrationReducer } from 'src/reducers/registration/'

describe('registration reducer', ()=> {
  it('should initialize correctly', () => {
    expect(registrationReducer(undefined, { type: '@INIT' })).toMatchSnapshot()
  })
})
