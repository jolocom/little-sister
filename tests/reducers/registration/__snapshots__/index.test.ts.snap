// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registration reducer should initialize correctly 1`] = `
Object {
  "loading": Immutable.Map {
    "loadingMsg": "Loading",
  },
}
`;
